### ISSUE TITLE
RyuJIT call optimization and aggressive inlining with known generic types 

### ISSUE THREAD

**Issue by redknightlois** [dotnet/coreclr#1532]
This probably will end up in the future releases wishlist, but it something that has been looking forward for a long time already.

Lets say that we have this code:

```csharp
       public class Executer<T> where T : ICalls
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            public void Execute(T instance)
            {
                instance.Execute();
            }
        }
```

And we have the following instances:

```csharp
        private readonly Executer<ClassCalls> _classCalls = new Executer<ClassCalls>();
        private readonly Executer<ICalls> _interfaceCalls = new Executer<ICalls>();
        private readonly Executer<SealedClassCalls> _sealedCalls = new Executer<SealedClassCalls>();
```

Now we would expect that the call for _classCalls.Execute(x) would be different than for _interfaceCalls(x). Apparently that is not the case, the JIT stops at the first level even if have the complete information to emit highly optimized code for that call-site.

Now, supposed the implementation is:
```csharp
        public class ClassCalls : ICalls
        {
            public static int i;

            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            public void Execute()
            {
                i = 0;
                i++;
            }
        } 
```

There is no way that the JIT would inline that code, even if for all purposes it is safe to do so.

The scenario for this pattern is pretty common in high performance code where the calls are very small, in tight loops but must be able to handle more than a single type... An example is a BitVector with variants for MemoryMappedBitVector, UnsafeBitVector, LongBitVector and so on. Operations tend to be very small and executed in very tight loops.

 Today we either need a different codepath for each one, or pay the call tax.

**Comment by mikedn** [comment]
I'm not quite sure what this has to do with generics, this looks more like a devirtualization problem.

**Comment by panost** [comment]
If I remember correctly, if ClassCalls was a struct and not a class, the Execute method would be inlined

**Comment by mikedn** [comment]
@panost Yes, in the case of structs the JIT usually does devirtualization. It's practically forced to do so, making an interface call on a value type would require boxing and you'd end up calling a method on a copy of the original value.

**Comment by redknightlois** [comment]
@mikedn yeah, but my benchmarks and the emitted assembly suggest that the JIT will perform some limited devirtualization if no generic type is involved. I agree that probably the topic could be changed to "RyuJIT support aggressive call devirtualization over constrained generic types" or something better if we can think a better one.

**Comment by mikedn** [comment]
> but my benchmarks and the emitted assembly suggest that the JIT will perform some limited devirtualization if no generic type is involved

Do you have some sample code?

**Comment by redknightlois** [comment]
@mikedn Sure. https://gist.github.com/redknightlois/5bafa47ee9835605da26

Just don't execute the naked call versions along with the generic ones (the difference between the count of instruction per each will screw the results --- probably as I am not counting properly the source instructions so I am not passing the right number to BenchmarkDotNet to do a proper adjustment). 

In there you will see that the timing for all the naked calls (sealed, unsealed and interface) have essentially the same cost. The assembly emitted for the 3 is identical as far as I remember. This suggest that some limited devirtualization is happening. 

@CarolEidt can you provide some insight here?

**Comment by mikedn** [comment]
`UseNakedInstanceCalls` and `UseNakedSealedCalls` do not contain any virtual/interface calls and generate identical code. `UseNakedInterfaceCalls` contains an interface call, at least in theory, be devirtualized. All the generic variants contain interface calls similar to `UseNakedInterfaceCalls`. They should generate the same code as `UseNakedInterfaceCalls` but there's some dead code that the JIT doesn't eliminate:

```
 sub         rsp,28h 
 mov         r11,qword ptr [rcx+30h]  ;dead
 mov         rcx,qword ptr [rcx+18h] 
 mov         r11d,dword ptr [r11]     ;dead
 mov         r11,7FFD68470048h 
 cmp         dword ptr [rcx],ecx 
 call        qword ptr [r11] 
 nop 
 add         rsp,28h 
 ret 
```

> In there you will see that the timing for all the naked calls (sealed, unsealed and interface) have essentially the same cost. The assembly emitted for the 3 is identical as far as I remember. This suggest that some limited devirtualization is happening.

I haven't measured the time but the naked interface variant certainly generates different code from the other 2 naked variants.

As for actually doing devirtualization in this case - it isn't that simple. For example, the call in `UseNakedInterfaceCalls` can only be devirtualized if the JIT observes that the `_nakedInterfaceCalls` is readonly (trivial) and initialized to an instance of `ClassCalls` (not that trivial as the initialization is done in the constructor, a different and unrelated method).

**Comment by redknightlois** [comment]
@mikedn And then I remember that I have RyuJIT disabled :)

These are the legacy JIT calls:

![devirtualization 1](https://cloud.githubusercontent.com/assets/1045411/9823852/43492c86-589f-11e5-9cb8-b7d526f7a2a6.png)
![devirtualization 2](https://cloud.githubusercontent.com/assets/1045411/9823853/4407a382-589f-11e5-891a-ef6b87952b9c.png)
![devirtualization 3](https://cloud.githubusercontent.com/assets/1045411/9823859/4f3d662e-589f-11e5-9e05-fe0c8b2f94bf.png)

Not a huge timing difference in between the alternatives.

**Comment by mikedn** [comment]
> These are the legacy JIT calls:

That's the 32 bit JIT, not the legacy (aka JIT64) JIT. Though on my machine JIT32 does inline the first two calls...

**Comment by redknightlois** [comment]
@mikedn I hate the "Prefer 32bits" option set by default of Visual Studio. See the edit.

**Comment by mikedn** [comment]
That `lea` is generated by the legacy JIT (JIT64) compiler, it's not generated by RyuJIT. Discussing the code generated by JIT64 isn't exactly useful.

**Comment by redknightlois** [comment]
@mikedn I know, that's why I said: "And then I remember that I have RyuJIT disabled :)" ... the limited devirtualization I've seen was JIT64 not RyuJIT making the whole argument moot. 

However, that doesn't invalidate the fact that some devirtualization used to happen, which apparently is not happening anymore.

**Comment by mikedn** [comment]
> the limited devirtualization I've seen was JIT64 not RyuJIT making the whole argument moot

But there's no kind of devirtualization going on in JIT64 either.

**Comment by redknightlois** [comment]
@mikedn OK now I see what you mean. For all uses and purposes those 2 call opcodes are equivalent. That interface call performance profile is the same of the register call for every processor upwards of Sandy Bridge (and maybe a couple of before). But, that's an artifact introduced by my code because I isolated the 3 calls in their own method. When called one after another (even creating the object in the line before) it can be seen that no devirtualization happens for the interface even if that would have been insanely safe. 

However, it can be argued that devirtualization of the type:
```csharp
ICall instance = new ClassCall();
instance.Execute();
```
could be done at the compiler level without much hassle. On the constrained generic types case that doesn't seem to be true.

**Comment by mikedn** [comment]
> However, it can be argued that devirtualization of the type: ... could be done at the compiler level without much hassle.

Yes, that's one case where devirtualization is possible. In itself it is a rather useless case as there's little reason to write such code to begin with (the only practical use for that kind of code is to access explicitly implement members). But such opportunities can show up in real code as the result of inlining of either the `Execute` call site (like it happens in your case, the generic `Execute` gets inlined so the callsite it contains can "see" the actual type assigned to `instance`) or of the instantiation site.

> On the constrained generic types case that doesn't seem to be true.

Generics don't play any part in this except for the fact that they introduce the interface call. For reference types your generic `Execute` method is no different from a non-generic method:

```C#
public class Executer
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static void Execute(ICalls instance)
    {
        instance.Execute();
    }
}
```

**Comment by benaadams** [comment]
Devirtualization of enumerators called via Interfaces back to structs would be nice...

**Comment by AndyAyersMS** [comment]
@redknightlois can you look this over and update if you still think there is anything actionable here, or close if not?

For generics instantiated over ref types we're unlikely to do devirtualization anytime soon, as the jit only sees the shared version. This might change down the road, if we somehow enabled unshared ref type instantiations or started looking into speculative devirtualization.

If the generic can get inlined into a context where the types are known then things open up a little and if the jit can put enough pieces together or see sealed types, it can do a lot of optimization.

**Comment by redknightlois** [comment]
@AndyAyersMS given that there are a few workarounds that could be found with sealed types and the actual solution for this is devirtualization of generic ref types I would say that the criteria for closing could be:
- [ ] Are the cases where it works documented?
- [ ] Are the workarounds documented?
- [ ] Are the limitations documented? Or, are there open issues that cover those cases?

If all are yes, I would say that this is done.
